README.md
The Early OOM Daemon
====================

[![Build Status](https://api.travis-ci.org/rfjakob/earlyoom.svg)](https://travis-ci.org/rfjakob/earlyoom)

The oom-killer generally has a bad reputation among Linux users. This may be
part of the reason Linux invokes it only when it has absolutely no other choice.
It will swap out the desktop environment, drop the whole page cache and empty
every buffer before it will ultimately kill a process. At least that's what I
think what it will do. I have yet to be patient enough to wait for it.

Instead of sitting in front of an unresponsive system, listening to the grinding
disk for minutes, I usually press the reset button and get back to what I was
doing quickly.

If you want to see what I mean, open something like
the Epic Citatel HTML5 Demo in a few Firefox windows (the
demo is now offline, it looked like
[this](https://blog.mozilla.org/blog/2014/03/12/mozilla-and-epic-preview-unreal-engine-4-running-in-firefox/)).
Save your work to disk beforehand, though.

The downside of the reset button is that it kills all processes, whereas it
would probably have been enough to kill a single one. This made people wonder
if the oom-killer could be configured to step in earlier: [superuser.com][2]
, [unix.stackexchange.com][3].

As it turns out, no, it can't. At least using the in-kernel oom killer.

In the user space however, we can do whatever we want.

What does it do
---------------
earlyoom checks the amount of available memory and (since version 0.5)
free swap 10 times a second. If both are below 10%, it will kill the
largest process. The percentage value is configurable via command line
arguments.

In the `free -m` output below, the available memory is 2170 MiB and
the free swap is 231 MiB.

                  total        used        free      shared  buff/cache   available
    Mem:           7842        4523         137         841        3182        2170
    Swap:          1023         792         231

Why is "available" memory checked as opposed to "free" memory?
On a healthy Linux system, "free" memory is supposed to be close to zero,
because Linux uses all available physical memory to cache disk access.
These caches can be dropped any time the memory is needed for something
else.

The "available" memory accounts for that. It sums up all memory that
is unused or can be freed immediately.

Note that you need a recent version of
`free` and Linux kernel 3.14+ to see the "available" column. If you have
a recent kernel, but an old version of `free`, you can get the value
from `cat /proc/meminfo | grep MemAvailable`.

When both your available memory and free swap drop below 10% of the total,
it will `kill -9` the process that uses the most memory in the opinion of
the kernel (`/proc/*/oom_score`). It can optionally (`-i` option) ignore
any positive adjustments set in `/proc/*/oom_score_adj` to protect innocent
victims (see below).

Why not trigger the kernel oom killer?
--------------------------------------
Earlyoom does not use `echo f > /proc/sysrq-trigger` because [the Chrome people made
their browser always be the first (innocent!) victim by setting `oom_score_adj`
very high]( https://code.google.com/p/chromium/issues/detail?id=333617).
Instead, earlyoom finds out itself by reading through `/proc/*/status`
(actually `/proc/*/statm`, which contains the same information but is easier to
parse programmatically).

Additionally, in recent kernels (tested on 4.0.5), triggering the kernel
oom killer manually may not work at all.
That is, it may only free some graphics
memory (that will be allocated immediately again) and not actually kill
any process. [Here](https://gist.github.com/rfjakob/346b7dc611fc3cdf4011) you
can see how this looks like on my machine (Intel integrated graphics).

How much memory does earlyoom use?
----------------------------------
About `0.6MB RSS`. All memory is locked using `mlockall()` to make sure
earlyoom does not slow down in low memory situations.

Download and compile
--------------------
Easy:

```bash
git clone https://github.com/rfjakob/earlyoom.git
cd earlyoom
make
sudo make install # Optional, if you want earlyoom to start
                  # automatically as a service (works on Fedora)
```

For Arch Linux, there's an [AUR package](https://aur.archlinux.org/packages/earlyoom/):
```bash
yaourt -S earlyoom
sudo systemctl enable earlyoom
sudo systemctl start earlyoom
```

Use
---
Just start the executable you have just compiled:

```bash
./earlyoom
```

It will inform you how much memory and swap you have, what the minimum
is, how much memory is available and how much swap is free.

```
earlyoom v0.10
mem total: 7842 MiB, min: 784 MiB (10 %)
swap total: 1023 MiB, min: 102 MiB (10 %)
mem avail: 5115 MiB (65 %), swap free: 1023 MiB (100 %)
mem avail: 5115 MiB (65 %), swap free: 1023 MiB (100 %)
mem avail: 5115 MiB (65 %), swap free: 1023 MiB (100 %)
[...]
```

If the values drop below the minimum, processes are killed until it
is above the minimum again. Every action is logged to stderr. If you are on
running earlyoom as a systemd service, you can view the last 10 lines
using

```bash
systemctl status earlyoom
```

### Notifications

The command-line flag `-n` enables notifications via `notify-send`. However, if earlyoom is being
run by a user other than the one running your desktop environment (e.g. if it's run as a service
or cron job) then `notify-send` will not work on its own, as DBUS, X, and/or display information
may required.

In this case, you can use `-N` to supply environment variables or another command. The exact value
will vary depending on your desktop environment, but the following command may work. `YOUR_USER`
should be replaced with output of `whoami` and `YOUR_USER_ID` with output of `echo $UID`. Your
`DISPLAY` value may also be different (check `echo $DISPLAY`).

```bash
earlyoom -N 'sudo -u YOUR_USER DISPLAY=:0 DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/YOUR_USER_ID/bus notify-send'
```

Other options are discussed in [this thread](https://unix.stackexchange.com/questions/111188/using-notify-send-with-cron).

Note that if you choose to use a command other than `notify-send`, it must support the same
arguments. Example invocation earlyoom uses:

```bash
NOTIFY_COMMAND -i dialog-warning 'notif title' 'notif text'
```

### Preferred Processes

The command-line flag `--prefer` specifies processes to prefer killing;
likewise, `--avoid` specifies
processes to avoid killing. The list of processes is specified by a regex expression.
For instance, to avoid having `foo` and `bar` be killed:

```bash
earlyoom --avoid '^(foo|bar)$'
```

The regex is matched against the basename of the process as shown
in `/proc/PID/stat`.

Command line options
--------------------
```
./earlyoom -h
earlyoom v0.12
Usage: earlyoom [OPTION]...

  -m PERCENT       set available memory minimum to PERCENT of total (default 10 %)
  -s PERCENT       set free swap minimum to PERCENT of total (default 10 %)
  -M SIZE          set available memory minimum to SIZE KiB
  -S SIZE          set free swap minimum to SIZE KiB
  -k               use kernel oom killer instead of own user-space implementation
  -i               user-space oom killer should ignore positive oom_score_adj values
  -n               enable notifications using "notify-send"
  -N COMMAND       enable notifications using COMMAND
  -d               enable debugging messages
  -v               print version information and exit
  -r INTERVAL      memory report interval in seconds (default 1), set to 0 to
                   disable completely
  -p               set niceness of earlyoom to -20 and oom_score_adj to -1000
  --prefer REGEX   prefer killing processes matching REGEX
  --avoid REGEX    avoid killing processes matching REGEX
  -h               this help text
```

Contribute
----------
Bug reports and pull requests are welcome via github. In particular, I am glad to
accept

* Use case reports and feedback

Changelog
---------
* v1.0, 2018-01-28: Add `--prefer` and `--avoid` options (@TomJohnZ)
* v0.12: Add `-M` and `-S` options (@nailgun); add man page, parameterize Makefile (@yangfl)
* v0.11: Fix undefined behavoir in get_entry_fatal (missing return, [commit](https://github.com/rfjakob/earlyoom/commit/9251d25618946723eb8a829404ebf1a65d99dbb0))
* v0.10: Allow to override Makefile's VERSION variable to make packaging easier,
  add `-v` command-line option
* v0.9: If oom_score of all processes is 0, use VmRss to find a victim
* v0.8: Use a guesstimate if the kernel does not provide MemAvailable
* v0.7: Select victim by oom_score instead of VmRSS, add options `-i` and `-d`
* v0.6: Add command-line options `-m`, `-s`, `-k`
* v0.5: Add swap support
* v0.4: Add SysV init script (thanks [@joeytwiddle](https://github.com/joeytwiddle)), use the new `MemAvailable` from `/proc/meminfo`
  (needs Linux 3.14+, [commit][4])
* v0.2: Add systemd unit file
* v0.1: Initial release

[1]: http://www.freelists.org/post/procps/library-properly-handle-memory-used-by-tmpfs
[2]: http://superuser.com/questions/406101/is-it-possible-to-make-the-oom-killer-intervent-earlier
[3]: http://unix.stackexchange.com/questions/38507/is-it-possible-to-trigger-oom-killer-on-forced-swapping
[4]: https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=34e431b0ae398fc54ea69ff85ec700722c9da773


debian/changelog
earlyoom (1.0) UNRELEASED; urgency=medium

  * Initial release.

 -- Mikko Rantalainen <mikko.rantalainen@iki.fi>  Sat, 03 Feb 2018 15:41:16 +0200


main.c
// SPDX-License-Identifier: MIT

/* Check available memory and swap in a loop and start killing
 * processes if they get too low */

#include <dirent.h>
#include <errno.h>
#include <getopt.h>
#include <regex.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/mman.h>
#include <sys/resource.h>
#include <unistd.h>

#include "kill.h"
#include "meminfo.h"

/* Arbitrary identifiers for long options that do not have a short
 * version */
#define LONG_OPT_PREFER 513
#define LONG_OPT_AVOID 514

int set_oom_score_adj(int);

int enable_debug = 0;

int main(int argc, char* argv[])
{
    int kernel_oom_killer = 0;
    unsigned long oom_cnt = 0;
    /* If the available memory goes below this percentage, we start killing
     * processes. 10 is a good start. */
    int mem_min_percent = 0, swap_min_percent = 0;
    long mem_min = 0, swap_min = 0; /* Same thing in KiB */
    int ignore_oom_score_adj = 0;
    char* notif_command = NULL;
    int report_interval = 1;
    int set_my_priority = 0;
    char* prefer_cmds = NULL;
    char* avoid_cmds = NULL;
    regex_t _prefer_regex;
    regex_t _avoid_regex;
    regex_t* prefer_regex = NULL;
    regex_t* avoid_regex = NULL;

    /* request line buffering for stdout - otherwise the output
     * may lag behind stderr */
    setlinebuf(stdout);

    fprintf(stderr, "earlyoom " VERSION "\n");

    if (chdir("/proc") != 0) {
        perror("Could not cd to /proc");
        exit(4);
    }

    DIR* procdir = opendir(".");
    if (procdir == NULL) {
        perror("Could not open /proc");
        exit(5);
    }

    int c;
    const char* short_opt = "m:s:M:S:kinN:dvr:ph";
    struct option long_opt[] = {
        { "prefer", required_argument, NULL, LONG_OPT_PREFER },
        { "avoid", required_argument, NULL, LONG_OPT_AVOID },
        { 0, required_argument, NULL, 0 }
    };

    while ((c = getopt_long(argc, argv, short_opt, long_opt, NULL)) != -1) {
        switch (c) {
        case -1: /* no more arguments */
        case 0: /* long option toggles */
            break;
        case 'm':
            mem_min_percent = strtol(optarg, NULL, 10);
            if (mem_min_percent <= 0) {
                fprintf(stderr, "-m: Invalid percentage\n");
                exit(15);
            }
            break;
        case 's':
            swap_min_percent = strtol(optarg, NULL, 10);
            if (swap_min_percent <= 0 || swap_min_percent > 100) {
                fprintf(stderr, "-s: Invalid percentage\n");
                exit(16);
            }
            break;
        case 'M':
            mem_min = strtol(optarg, NULL, 10);
            if (mem_min <= 0) {
                fprintf(stderr, "-M: Invalid KiB value\n");
                exit(15);
            }
            break;
        case 'S':
            swap_min = strtol(optarg, NULL, 10);
            if (swap_min <= 0) {
                fprintf(stderr, "-S: Invalid KiB value\n");
                exit(16);
            }
            break;
        case 'k':
            kernel_oom_killer = 1;
            fprintf(stderr, "Using kernel oom killer\n");
            break;
        case 'i':
            ignore_oom_score_adj = 1;
            break;
        case 'n':
            notif_command = "notify-send";
            break;
        case 'N':
            notif_command = optarg;
            break;
        case 'd':
            enable_debug = 1;
            break;
        case 'v':
            // The version has already been printed above
            exit(0);
        case 'r':
            report_interval = strtol(optarg, NULL, 10);
            if (report_interval < 0) {
                fprintf(stderr, "-r: Invalid interval\n");
                exit(14);
            }
            break;
        case 'p':
            set_my_priority = 1;
            break;
        case LONG_OPT_PREFER:
            prefer_cmds = optarg;
            break;
        case LONG_OPT_AVOID:
            avoid_cmds = optarg;
            break;
        case 'h':
            fprintf(stderr,
                "Usage: earlyoom [OPTION]...\n"
                "\n"
                "  -m PERCENT       set available memory minimum to PERCENT of total (default 10 %%)\n"
                "  -s PERCENT       set free swap minimum to PERCENT of total (default 10 %%)\n"
                "  -M SIZE          set available memory minimum to SIZE KiB\n"
                "  -S SIZE          set free swap minimum to SIZE KiB\n"
                "  -k               use kernel oom killer instead of own user-space implementation\n"
                "  -i               user-space oom killer should ignore positive oom_score_adj values\n"
                "  -n               enable notifications using \"notify-send\"\n"
                "  -N COMMAND       enable notifications using COMMAND\n"
                "  -d               enable debugging messages\n"
                "  -v               print version information and exit\n"
                "  -r INTERVAL      memory report interval in seconds (default 1), set to 0 to\n"
                "                   disable completely\n"
                "  -p               set niceness of earlyoom to -20 and oom_score_adj to -1000\n"
                "  --prefer REGEX   prefer killing processes matching REGEX\n"
                "  --avoid REGEX    avoid killing processes matching REGEX\n"
                "  -h               this help text\n");
            exit(1);
        case '?':
            exit(13);
        }
    }

    if (mem_min_percent && mem_min) {
        fprintf(stderr, "Can't use -m with -M\n");
        exit(2);
    }

    if (swap_min_percent && swap_min) {
        fprintf(stderr, "Can't use -s with -S\n");
        exit(2);
    }

    if (kernel_oom_killer && ignore_oom_score_adj) {
        fprintf(stderr, "Kernel oom killer does not support -i\n");
        exit(2);
    }

    if (kernel_oom_killer && (prefer_cmds || avoid_cmds)) {
        fprintf(stderr, "Kernel oom killer does not support --prefer/--avoid\n");
        exit(2);
    }

    if (prefer_cmds) {
        prefer_regex = &_prefer_regex;
        if (regcomp(prefer_regex, prefer_cmds, REG_EXTENDED | REG_NOSUB) != 0) {
            fprintf(stderr, "Could not compile regexp: %s\n", prefer_cmds);
            exit(6);
        }
    }

    if (avoid_cmds) {
        avoid_regex = &_avoid_regex;
        if (regcomp(avoid_regex, avoid_cmds, REG_EXTENDED | REG_NOSUB) != 0) {
            fprintf(stderr, "Could not compile regexp: %s\n", avoid_cmds);
            exit(6);
        }
    }

    struct meminfo m = parse_meminfo();

    if (mem_min) {
        mem_min_percent = 100 * mem_min / m.MemTotal;
    } else {
        if (!mem_min_percent) {
            mem_min_percent = 10;
        }
        mem_min = m.MemTotal * mem_min_percent / 100;
    }

    if (swap_min) {
        swap_min_percent = 100 * swap_min / m.SwapTotal;
    } else {
        if (!swap_min_percent) {
            swap_min_percent = 10;
        }
        swap_min = m.SwapTotal * swap_min_percent / 100;
    }

    fprintf(stderr, "mem total: %lu MiB, min: %lu MiB (%d %%)\n",
        m.MemTotal / 1024, mem_min / 1024, mem_min_percent);
    fprintf(stderr, "swap total: %lu MiB, min: %lu MiB (%d %%)\n",
        m.SwapTotal / 1024, swap_min / 1024, swap_min_percent);

    if (notif_command)
        fprintf(stderr, "notifications enabled using command: %s\n", notif_command);

    /* Dry-run oom kill to make sure stack grows to maximum size before
     * calling mlockall()
     */
    handle_oom(procdir, 0, kernel_oom_killer, ignore_oom_score_adj,
        notif_command, prefer_regex, avoid_regex);

    if (mlockall(MCL_CURRENT | MCL_FUTURE) != 0)
        perror("Could not lock memory - continuing anyway");

    if (set_my_priority) {
        if (setpriority(PRIO_PROCESS, 0, -20) != 0)
            perror("Could not set priority - continuing anyway");

        if (set_oom_score_adj(-1000) != 0)
            perror("Could not set oom_score_adj to -1000 for earlyoom process - continuing anyway");
    }

    c = 1; // Start at 1 so we do not print another status line immediately
    report_interval = report_interval * 10; // convert seconds to tenth of second
    while (1) {
        m = parse_meminfo();

        if (report_interval && c % report_interval == 0) {
            int swap_free_percent = 0;
            if (m.SwapTotal > 0)
                swap_free_percent = m.SwapFree * 100 / m.SwapTotal;

            printf("mem avail: %lu MiB (%ld %%), swap free: %lu MiB (%d %%)\n",
                m.MemAvailable / 1024, m.MemAvailable * 100 / m.MemTotal,
                m.SwapFree / 1024, swap_free_percent);
            c = 0;
        }
        c++;

        if (m.MemAvailable <= mem_min && m.SwapFree <= swap_min) {
            fprintf(stderr, "Out of memory! avail: %lu MiB < min: %lu MiB\n",
                m.MemAvailable / 1024, mem_min / 1024);
            handle_oom(procdir, 9, kernel_oom_killer, ignore_oom_score_adj,
                notif_command, prefer_regex, avoid_regex);
            oom_cnt++;
        }

        usleep(100000); // 100ms
    }

    return 0;
}

int set_oom_score_adj(int oom_score_adj)
{
    char buf[256];
    pid_t pid = getpid();

    snprintf(buf, sizeof(buf), "%d/oom_score_adj", pid);
    FILE* f = fopen(buf, "w");
    if (f == NULL) {
        return -1;
    }

    fprintf(f, "%d", oom_score_adj);
    fclose(f);

    return 0;
}
